File: src/server/modules/ServerComponents/Grippable.lua
local CollectionService = game:GetService("CollectionService")
local Players = game:GetService("Players")
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local RunService = game:GetService("RunService")

local Knit = require(ReplicatedStorage.Packages.Knit)
local Trove = require(ReplicatedStorage.Packages.Trove)
local Component = require(ReplicatedStorage.Packages.Component)
local Logger = require(ReplicatedStorage.Extensions.Logger)

local HealthService, RagdollService

local Grippable = Component.new {
	Tag = "Grippable";
	Extensions = {
		Logger,
	};
}

function Grippable:Construct()
	self._trove = Trove.new()

	self.gettingGripped = false
end

function Grippable:Grip(player: Player, raycastResult: RaycastResult)
	local gripper: Model = player.Character
	print(gripper.Name, "gripping", self.Instance.Name)

	self._gripTrove = self._trove:Extend()
	self._gripTrove:Connect(gripper:GetAttributeChangedSignal("Ragdolled"), function()
		print(gripper.Name, "ragdolled; releasing", self.Instance.Name)
		self:Release(player)
	end)
	self._gripTrove:Connect(gripper:GetAttributeChangedSignal("Dead"), function()
		print(self.Instance.Name, "died; releasing", gripper.Name)
		self:Release(player)
	end)

	self.gettingGripped = true
	local grippee: Model = self.Instance

	grippee:SetAttribute("GettingGripped", true)
	gripper:SetAttribute("Gripping", grippee.Name)

	local grippeeHumanoid = grippee:FindFirstChildWhichIsA("Humanoid")
	local gripperHumanoid = gripper:FindFirstChildWhichIsA("Humanoid")
	if not grippeeHumanoid then error("nil humanoid(s)") end
	CollectionService:RemoveTag(grippeeHumanoid, "Burning")
	grippeeHumanoid.AutoRotate = false
	gripperHumanoid.AutoRotate = false

	local grippeeAnimator: Animator = grippeeHumanoid.Animator
	local gripperAnimator: Animator = gripperHumanoid.Animator

	local hitSound: Sound = grippee.Torso:FindFirstChild("PunchHit") or self._trove:Clone(ReplicatedStorage.Effects.Sounds.PunchHit)
	hitSound.Parent = grippee.Torso
	local hitParticle: ParticleEmitter = grippee.Head:FindFirstChild("PunchEmit") or self._trove:Clone(ReplicatedStorage.Effects.Visuals.PunchEmit)
	hitParticle.Parent = grippee.Head

	self.Gripped = grippeeAnimator:LoadAnimation(ReplicatedStorage.Animations.Combat.Gripped)
	self._gripTrove:Connect(self.Gripped.KeyframeReached, function(keyframeName)
		if keyframeName == "Resting" then
			self.Gripped:AdjustSpeed(0)
		end
	end)

	self.Gripping = gripperAnimator:LoadAnimation(ReplicatedStorage.Animations.Combat.Gripping)
	self._gripTrove:Connect(self.Gripping.KeyframeReached, function(keyframeName)
		if keyframeName == "FinalHit" then
			self.Gripped:Play()
			hitSound:Play()
			hitParticle:Emit(1)
			self._ticks += 1
			if self._ticks >= 5 then
				print("grip complete")

				grippeeHumanoid.Health = 0
				local grippedPlayer = Players:GetPlayerFromCharacter(grippee)
				if grippedPlayer ~= nil then HealthService.Kill(grippedPlayer) end
				RagdollService.Pop(grippee)

				gripper.HumanoidRootPart.RootJoint.Enabled = false
				RunService.Stepped:Wait()
				gripper.HumanoidRootPart.RootJoint.Enabled = true

				hitSound.Ended:Wait()
				self:Release(player)
			end
		end
	end)

	self._ticks = 0
	self.Gripping:Play()
	self.Gripped:Play(0, 1, 0)

	grippee.PrimaryPart.Anchored = true
	local pos = raycastResult.Position + Vector3.yAxis*0.5
	local zRotation = grippee.PrimaryPart.Orientation.Z
	grippee:PivotTo(CFrame.lookAt(pos, pos + Vector3.yAxis))
	grippee:PivotTo(grippee:GetPivot() * CFrame.Angles(0, 0, math.rad(zRotation)))

	for _,joint in grippee.Torso:GetChildren() do
		if joint:IsA("Motor6D") then
			joint.Enabled = true
		end
	end

	local positionOffset = Vector3.new(0,2.5,0) + -grippee.PrimaryPart.CFrame.UpVector * 3
	local rotationOffset = CFrame.Angles(math.rad(90), math.rad(0), math.rad(180))
	gripper.PrimaryPart.Anchored = true
	gripper:PivotTo(grippee:GetPivot() * rotationOffset + positionOffset)
end

function Grippable:Release(player: Player)
	self._gripTrove:Destroy()

	self.gettingGripped = false
	self._ticks = 0

	self.Gripping:Stop()
	self.Gripped:Stop()
	self.Gripping = nil
	self.Gripped = nil

	local grippee: Model = self.Instance
	local gripper: Model = player.Character
	if grippee.Parent == workspace.Dead then return end

	local humanoid = gripper:FindFirstChildWhichIsA("Humanoid")
	humanoid.AutoRotate = true
	for _,joint in grippee.Torso:GetChildren() do
		if joint:IsA("Motor6D") then
			joint.Enabled = false
		end
	end

	grippee:SetAttribute("GettingGripped", nil)
	gripper:SetAttribute("Gripping", nil)

	grippee.PrimaryPart.Anchored = false
	gripper.PrimaryPart.Anchored = false
end

--function Grippable:OnTriggered(player: Player)
--	if player.Character == self.Instance then return end

--	if not self.isCarried then
--		self:PickUp(player)
--	else
--		self:Release(player)
--	end
--end

--function Grippable:OnKnocked(...)
--	self.Prompt.Enabled = self.Instance:GetAttribute("Knocked") or false
--end

function Grippable:Start()
	HealthService = Knit.GetService("HealthService")
	RagdollService = Knit.GetService("RagdollService")

	--self._trove:Connect(self.Prompt.Triggered, function(...) self:OnTriggered(...) end)
	--self._trove:Connect(self.Instance:GetAttributeChangedSignal("Knocked"), function(...) self:OnKnocked(...) end)
end

function Grippable:Stop()
	self._trove:Destroy()
end

return Grippable

File: src/server/components/character/grippable.ts
import { Component, Components } from "@flamework/components";
import { OnStart } from "@flamework/core";
import { Workspace } from "@rbxts/services";
import { ANIMATIONS, SFX, VFX } from "shared/constants";
import { Inject } from "shared/inject";
import { KeyInteractable } from "../interactable/key-interactable";
import { CharacterServer } from "./character-server";
import { RagdollServer } from "./ragdoll-server";

interface Attributes {
	gettingGripped: boolean;
}

const GETTING_GRIPPED_ANIMATION = ANIMATIONS.Combat.Gripped;
const GRIPPING_ANIMATION = ANIMATIONS.Combat.Gripping;

@Component({
	tag: "Grippable",
	defaults: {
		gettingGripped: false,
	},
})
export class Grippable
	extends KeyInteractable<Attributes, Model>
	implements OnStart
{
	private gripTrove = this.trove.extend();

	@Inject
	private components!: Components;

	public constructor(
		private character: CharacterServer,
		private ragdoll: RagdollServer,
	) {
		super();
	}

	public override onStart(): void {
		super.onStart();

		this.toggle(false);
		this.setKey(Enum.KeyCode.B);
		this.trove.add(
			this.ragdoll.onAttributeChanged("isRagdolled", (newValue) => {
				this.toggle(newValue);
			}),
		);
	}

	public override interact(player: Player): void {
		if (!player.Character) return;
		const characterComponent =
			this.components.getComponent<CharacterServer>(player.Character);
		if (!characterComponent) return;
		print(this.attributes.gettingGripped);
		this.attributes.gettingGripped
			? this.release(characterComponent)
			: this.grip(characterComponent);
	}

	public grip(gripper: CharacterServer): void {
		const floorCheck = Workspace.Raycast(
			this.character.getHumanoidRootPart().Position,
			Vector3.yAxis.mul(-4),
		);

		if (!floorCheck) error(`${this.instance.Name} is mid-air`);

		this.attributes.gettingGripped = true;

		const ragdoll = this.components.getComponent<RagdollServer>(
			gripper.instance,
		);
		if (ragdoll) {
			this.gripTrove.add(
				ragdoll.onAttributeChanged("isRagdolled", (newValue) => {
					if (newValue) this.release(gripper);
				}),
			);
		}

		// detect gripper death/disconnect

		this.instance.RemoveTag("Burning");

		const gripperHumanoid = gripper.getHumanoid();
		const humanoid = this.character.getHumanoid();

		humanoid.AutoRotate = false;
		gripperHumanoid.AutoRotate = false;

		const torso = this.character.getTorso();
		const head = this.character.getHead();
		const hitSound = this.gripTrove.clone(SFX.PunchHit);
		const hitParticle = this.gripTrove.clone(VFX.PunchEmit);
		hitSound.Parent = torso;
		hitParticle.Parent = head;

		const grippedAnimationTrack = this.character
			.getAnimator()
			.LoadAnimation(GETTING_GRIPPED_ANIMATION);
		this.gripTrove.add(() => grippedAnimationTrack.Stop());
		this.gripTrove.connect(
			grippedAnimationTrack.KeyframeReached,
			(keyframeName) => {
				if (keyframeName === "Resting")
					grippedAnimationTrack.AdjustSpeed(0);
			},
		);

		let ticks = 0;
		const grippingAnimationTrack = gripper
			.getAnimator()
			.LoadAnimation(GRIPPING_ANIMATION);
		this.gripTrove.add(() => grippingAnimationTrack.Stop());
		this.gripTrove.connect(
			grippingAnimationTrack.KeyframeReached,
			(keyframeName) => {
				if (keyframeName !== "FinalHit") return;

				grippingAnimationTrack.Play();
				hitSound.Play();
				hitParticle.Emit(1);

				ticks++;
				if (ticks < 5) return;

				this.character.kill();

				hitSound.Ended.Wait();
				this.release(gripper);
			},
		);

		grippingAnimationTrack.Play();

		// gripee may need to have motor6ds reenabled because of ragdoll
		grippedAnimationTrack.Play(0, 1, 0);

		this.snapToGround(floorCheck, gripper, this.character);
	}

	public release(gripper: CharacterServer): void {
		this.attributes.gettingGripped = false;

		this.gripTrove.clean();

		gripper.getHumanoid().AutoRotate = true;
		gripper.getHumanoidRootPart().Anchored = false;
		this.character.getHumanoidRootPart().Anchored = false;

		// redisable motor6ds if ragdolled
	}

	private snapToGround(
		floorCheck: RaycastResult,
		gripper: CharacterServer,
		grippee: CharacterServer,
	) {
		const gripperRootPart = gripper.getHumanoidRootPart();
		const grippeeRootPart = grippee.getHumanoidRootPart();
		const zRotation = grippeeRootPart.Orientation.Z;

		const goalPos = floorCheck.Position.add(Vector3.yAxis.mul(0.5));

		grippeeRootPart.Anchored = true;
		grippee.instance.PivotTo(
			CFrame.lookAt(goalPos, goalPos.add(Vector3.yAxis)),
		);
		const grippeePivot = grippee.instance.GetPivot();
		grippee.instance.PivotTo(
			grippeePivot.mul(CFrame.Angles(0, 0, zRotation)),
		);

		const gripperPosOffset = Vector3.yAxis
			.mul(2.5)
			.add(grippeeRootPart.CFrame.UpVector.mul(3));
		const gripperRotOffset = CFrame.Angles(math.rad(90), 0, math.rad(180));

		gripperRootPart.Anchored = true;
		gripper.instance.PivotTo(
			grippeePivot.mul(gripperRotOffset).add(gripperPosOffset),
		);
	}
}
